src/main/java/spoon/reflect/visitor/chain/CtQueryListener.java

```diff
@@ -0,0 +1,41 @@
+/**
+ * Copyright (C) 2006-2017 INRIA and contributors
+ * Spoon - http://spoon.gforge.inria.fr/
+ *
+ * This software is governed by the CeCILL-C License under French law and
+ * abiding by the rules of distribution of free software. You can use, modify
+ * and/or redistribute the software under the terms of the CeCILL-C license as
+ * circulated by CEA, CNRS and INRIA at http://www.cecill.info.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the CeCILL-C License for more details.
+ *
+ * The fact that you are presently reading this means that you have had
+ * knowledge of the CeCILL-C license and that you accept its terms.
+ */
+package spoon.reflect.visitor.chain;
+
+/**
+ * Listens for entering/exiting of query step
+ */
+public interface CtQueryListener {
+	/**
+	 * Called when new object enters query step, before the object is processed by mapping function of that step.
+	 * <br>
+	 * The returning false or throwing an exception causes that processing of input object and all it's children is skipped.
+	 * The returning true causes that input object is normally processed, and exit is called for that object.
+	 * @param element the processed element
+	 * @param context context the query execution context
+	 * @return true to continue processing this step and it's children,
+	 * false to skip this step and it's children.
+	 */
+	boolean enter(Object element, CtQuery context);
+	/**
+	 * The exit is called always only after `enter` returns true and input element and all children are processed or they throw an exception.
+	 * @param element the processed element
+	 * @param context the query execution context
+	 * @param exception null or an exception if the processing of the mapping function failed

```

Comment:
actually it is used like this:
```java
listener.enter(item, query);
try {
 ... process children
} catch (Throwable e) {
  listener.exit(item, query, e);
  throw e;
}
listener.exit(item, query, null);
```
so the exit is called always, and you can detect in exit whether it failed or not by checking exception parameter.

Optionally we might implement it like this
```java
listener.enter(item, query);
try {
 ... process children
} catch (Throwable e) {
  listener.failureExit(item, query, e);
  throw e;
}
listener.exit(item, query);
```
where `failureExit` method might added to the API later, after somebody needs it.
In such case we should provide an default implementation of listener, so the clients can extend that default implementation and will be not forced to change their code later, after we add method failureExit into API.